### PR TITLE
Refactor App data collection

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

To separate the collection of application data from its serialisation, supporting the Unity C# layer changes.

## Changeset

- Remove all methods from `AppData` and `AppDataSummary` which aren't accessors or serialisers
- Create `AppDataCollector`, which collects data for `AppData`, `AppDataSummary`, and app `MetaData`.
- Reduce number of packageManager lookups + cache other suitable values in `AppDataCollector` compared to `AppData`
- Generate a new `AppData` instance for each error report.
- Remove error's dependency on `AppData#getActiveScreenClass()`
- Add `Client#populateAppMetaData(MetaData metaData)` method for accessing extra app data as internal API

## Tests

Ran mazerunner and existing unit test suite. Added unit tests for new methods.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: sdk/src/main/java/com/bugsnag/android/AppData.java [hunk 1]
```diff
@@ -15,15 +10,7 @@
...
-    private static final long startTimeMs = SystemClock.elapsedRealtime();
-
-    @Nullable
-    final String appName;
-
-    @NonNull
-    private final Context appContext;
-    private final SessionTracker sessionTracker;
-
+    @SuppressWarnings("NullableProblems") // initialised after construction
```
Comment (fractalwrench) [inline sdk/src/main/java/com/bugsnag/android/AppData.java]: Warnings are suppressed as we're anticipating that this field will always be set to a non-null value after construction. This will allow users to access the value in a Kotlin-esque way, i.e.:

`appData.packageName`

Rather than requiring the safe call operator each time to unwrap the optional:

`appData?.packageName`
